Add docs for transparent proxy mode and config (#10038)
hashicorp/consul: website/content/docs/connect/transparent-proxy-k8s.mdx

> ---

Reviewer: This is a copy of transparent-proxy.mdx, since I wasn't able to have 2 pages on docs-nav-data.json point to the same mdx file for some reason. Posting here in case someone has an idea for a way to not just copy this file :)

Reviewer 2: I've added href field instead, that way we can just use one page!

registration](https://www.consul.io/docs/connect/registration/service-registration) on VMs and via the
[annotation](https://www.consul.io/docs/k8s/connect#consul-hashicorp-com-connect-service-upstreams) in Kubernetes.

To support transparent proxying, Consul now supports a command `consul connect redirect-traffic` to redirect traffic
through an inbound and outbound listener on Envoy. It also watches Service Intentions and configures the Envoy proxy
with the appropriate upstream IPs. In Consul on Kubernetes, the traffic redirection command is automatically set up via
an init container.

## Prerequisites

Transparent proxy requires Consul >= `1.10.0`.

### Kubernetes

* To use transparent proxy on Kubernetes, Consul-helm >= `0.32.0` and Consul-k8s >= `0.26.0` are required in addition to
the Consul version requirements.
* Service Intentions should be set up to allow intended services to connect to each other.

The Kubernetes integration takes care of registering Kubernetes services with Consul, injecting a sidecar proxy, and
enabling traffic redirection.

### VMs

* For a service on a VM to be a part of the service mesh, it needs to run a Connect Envoy proxy.

> * The `consul connect redirect-traffic` command needs to be run on the VM to set it up to redirect all inbound and

Reviewer: great idea!

transparent_proxy {
  outbound_listener_port = 15001
}
```
```
# Proxy service registrations apply to a single proxy instance.
name = "web-sidecar-proxy"
kind = "connect-proxy"
proxy {
  mode = "transparent"
  transparent_proxy {
    outbound_listener_port = 15001
  }
  destination_service_name = "web"
  local_service_port = 8080
}
port = 20000
```

Similar to `mesh_gateway.mode`, the new proxy mode will have the following string values:
* "" - The empty string represents the default value for the feature, and allows for the mode to be overridden by
central configuration, like “service-defaults”.
* "direct" - Explicitly disables configuring Envoy as a transparent proxy, falling back to only configuring explicit upstreams.
* "transparent" - Explicitly enables configuring Envoy as a transparent proxy.

Additionally, the new Cluster config entry is scoped to the set of federated Consul datacenters and can be used to allow or block
traffic to external destinations. This example shows blocking traffic to external destinations (outside of Consul's catalog):

```
kind = "cluster"
name = "cluster"

transparent_proxy {
  catalog_destinations_only = true
}
```

## Known Limitations

> * For services on VMs, transparent proxy is only supported if there is one service per VM. This is because the traffic

Reviewer 3: ```suggestion
* For services on VMs, transparent proxy only supports one service per VM, or one service per network namespace. This is because the traffic
```

for Connect applications, i.e:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: sample-app
spec:
  selector:
    app: sample-app
```

In the example above, if another service wants to reach `sample-app` via transparent proxying, it just needs a
[ServiceIntention](https://www.consul.io/docs/connect/config-entries/service-intentions) allowing it to talk to
`sample-app` and then it can dial `sample-app.svc.cluster.local`, using
[KubeDNS](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/).

### VMs
To use transparent proxy on VMs, the service needs to be registered with Consul and a connect proxy needs to be added to
the mesh on the VM. Then, traffic redirection rules need to be set up to direct inbound and outbound traffic through the
Envoy connect proxy. Then, to enable transparent proxy mode to reach this service, you can set apply a service defaults
config entry to configure the mode to be transparent as shown above in the [Configuration section](#configuration).

Now, once Service Intentions are set up, other services can reach this service's address via an address known to Consul,
and the traffic will go through the proxy.

> ~> **Note** Only one service is supported per VM. See [Known Limitations](#known-limitations)

Reviewer 3: ```suggestion
~> **Note** Only one service is supported per VM, or per network namespace. See [Known Limitations](#known-limitations)
```
